### PR TITLE
Fix FFI-unwind unsoundness with mixed panic mode

### LINKED ISSUE #96926: Mixing crates using different panic modes may introduce UB if a non-Rust unwind "escapes" into panic=abort crate via "C-unwind"
This is a case of undefined behavior around unwinding that will not be fixed by the stabilization of the `"C-unwind"` ABI.

Suppose there exists a C++ library with this function:

```c++
void cpp_unwind() {
    throw "gotcha";
}
```

If this function is compiled as part of a crate with `panic=unwind`:

```rust
extern "C-unwind" {
    cpp_unwind()
}

pub fn unwinds() { unsafe { cpp_unwind() } }
```

...and then called from a crate compiled with `panic=abort`, the behavior is undefined, because the calling function will be compiled with the assumption that unwinding is impossible.

Note:

* If `cpp_unwind` is declared with `extern "C"`, the behavior will be well-defined: the runtime will abort even if `panic=unwind` is used.
* Cargo does not support mixing panic modes like this, so this situation is only possible by invoking `rustc` directly (or via a different build system).
* The unwind must originate outside of Rust, because only one `panic!()` implementation will be linked in the final binary, so using `panic=abort` will prevent `panic!()` from unwinding the stack in the first place.

### PR DESCRIPTION
**Pull request by nbdd0121**
UB maybe introduced when an FFI exception happens in a `C-unwind` foreign function and it propagates through a crate compiled with `-C panic=unwind` into a crate compiled with `-C panic=abort` (#96926).

To prevent this unsoundness from happening, we will disallow a crate compiled with `-C panic=unwind` to be linked into `panic-abort` *if* it contains a call to `C-unwind` foreign function or function pointer. If no such call exists, then we continue to allow such mixed panic mode linking because it's sound (and stable). In fact we still need the ability to do mixed panic mode linking for std, because we only compile std once with `-C panic=unwind` and link it regardless panic strategy.

For libraries that wish to remain compile-once-and-linkable-to-both-panic-runtimes, a `ffi_unwind_calls` lint is added (gated under `c_unwind` feature gate) to flag any FFI unwind calls that will cause the linkable panic runtime be restricted.

In summary:
```rust
#![warn(ffi_unwind_calls)]

mod foo {
    #[no_mangle]
    pub extern "C-unwind" fn foo() {}
}

extern "C-unwind" {
    fn foo();
}

fn main() {
    // Call to Rust function is fine regardless ABI.
    foo::foo();
    // Call to foreign function, will cause the crate to be unlinkable to panic-abort if compiled with `-Cpanic=unwind`.
    unsafe { foo(); }
    //~^ WARNING call to foreign function with FFI-unwind ABI
    let ptr: extern "C-unwind" fn() = foo::foo;
    // Call to function pointer, will cause the crate to be unlinkable to panic-abort if compiled with `-Cpanic=unwind`.
    ptr();
    //~^ WARNING call to function pointer with FFI-unwind ABI
}
```

Fix #96926

@rustbot label: T-compiler F-c_unwind

### CODE REVIEW

File: compiler/rustc_metadata/src/rmeta/mod.rs [hunk 1]
```diff
@@ -198,7 +198,7 @@ crate struct CrateRoot<'tcx> {
     extra_filename: String,
     hash: Svh,
     stable_crate_id: StableCrateId,
-    panic_strategy: PanicStrategy,
+    panic_strategy: Option<PanicStrategy>,
```
Comment (BatmanAoD) [inline compiler/rustc_metadata/src/rmeta/mod.rs:201]: What would cause this to be `None`?

Comment (nbdd0121) [reply compiler/rustc_metadata/src/rmeta/mod.rs:201]: If the crate is compiled with `-C panic=unwind` and contains no FFI unwind calls. I.e. it does not require a particular panic runtime to be linked.

Comment (BatmanAoD) [reply compiler/rustc_metadata/src/rmeta/mod.rs:201]: Got it, thanks!

**Comment by compiler-errors** [comment]
not sure if i'm the best person to review this

r? rust-lang/compiler

**Comment by RalfJung** [comment]
> // Call to Rust function is fine regardless ABI.

Why is that okay?
I first thought the check would be "reject if there are any calls with C-unwind ABI", but clearly what actually happens is a bit more subtle, so I want to make sure we document precisely why this is okay. :)

(And that documentation should be somewhere more permanent that this PR, such as in the code and maybe even ameded to the RFC or added to the reference or so.)

File: compiler/rustc_metadata/src/rmeta/mod.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_metadata/src/rmeta/mod.rs:201]: That seems a bit odd? The panic strategy still makes a different for other things, doesn't it?

File: compiler/rustc_mir_transform/src/ffi_unwind_calls.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+// Check if the body of this def_id can possibly leak a foreign unwind into Rust code.
+fn has_ffi_unwind_calls(tcx: TyCtxt<'_>, local_def_id: LocalDefId) -> bool {
+    debug!("has_ffi_unwind_calls({local_def_id:?})");
+
+    // Only perform check on functions because constants cannot call FFI functions.
+    let def_id = local_def_id.to_def_id();
+    let kind = tcx.def_kind(def_id);
+    let is_function = match kind {
+        DefKind::Fn | DefKind::AssocFn | DefKind::Ctor(..) => true,
+        _ => tcx.is_closure(def_id),
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/ffi_unwind_calls.rs:49]: Is this a dangerous fallback `_` that could become wrong if more kinds are added in the future?

File: compiler/rustc_metadata/src/dependency_format.rs [hunk 3]
```diff
@@ -397,18 +401,14 @@ fn verify_ok(tcx: TyCtxt<'_>, list: &[Linkage]) {
             if let Linkage::NotLinked = *linkage {
                 continue;
             }
-            if desired_strategy == PanicStrategy::Abort {
-                continue;
-            }
```
Comment (RalfJung) [inline compiler/rustc_metadata/src/dependency_format.rs:402]: This is a key part of the PR, right? That's where previously we skipped a check that we do not skip any more now.

**Comment by RalfJung** [comment]
> Why is that okay?

Trying to answer my own question... Is it correct that this PR ensures the following property?

If anywhere in the crate graph is a panic=abort crate, then all call sites where foreign exceptions could enter the Rust part of the program are in a panic=abort crate, and hence will lead to immediate abort.
These call sites are C-unwind call sites to FFI functions or function pointers -- a call site calling a Rust function cannot be where anything enters Rust.

File: library/std/src/lib.rs [hunk 4]
```diff
@@ -210,6 +210,8 @@
 #![allow(unused_lifetimes)]
 // Tell the compiler to link to either panic_abort or panic_unwind
 #![needs_panic_runtime]
+// Ensure that std can be linked against panic_abort despite compiled with `-C panic=unwind`
+#![cfg_attr(not(bootstrap), deny(ffi_unwind_calls))]
```
Comment (RalfJung) [inline library/std/src/lib.rs:214]: Don't `core` and `alloc` need the same?

**Comment by RalfJung** [comment]
Cc @Amanieu

File: compiler/rustc_metadata/src/rmeta/mod.rs (hunk 1 above)
Comment (Amanieu) [reply compiler/rustc_metadata/src/rmeta/mod.rs:201]: This field should definitely be renamed (perhaps `requires_panic_runtime`?), it is very confusing at the moment since "panic strategy" refers to how the current crate was compiled.

File: compiler/rustc_interface/src/passes.rs [hunk 5]
```diff
@@ -954,6 +954,7 @@ fn analysis(tcx: TyCtxt<'_>, (): ()) -> Result<()> {
             if !tcx.sess.opts.debugging_opts.thir_unsafeck {
                 rustc_mir_transform::check_unsafety::check_unsafety(tcx, def_id);
             }
+            tcx.ensure().has_ffi_unwind_calls(def_id);
```
Comment (Amanieu) [inline compiler/rustc_interface/src/passes.rs:957]: This seems to be called twice: here and in `mir_const`. Is this intentional?

File: compiler/rustc_metadata/src/rmeta/mod.rs (hunk 1 above)
Comment (bjorn3) [reply compiler/rustc_metadata/src/rmeta/mod.rs:201]: `#![no_std]` code doesn't require a panic runtime, but could still require a particular panic strategy for any dependent crates. For example a `-Cpanic=abort` no std crate requires all dependent crates to be compiled with `-Cpanic=abort` too, even though using `#[panic_handler]` rather than linking a panic runtime would work.

Comment (nbdd0121) [reply compiler/rustc_metadata/src/rmeta/mod.rs:201]: This field and related query (`panic_strategy(_: CrateNum)`) is only used for checking compatibility. I could rename this to `required_panic_strategy`.

File: compiler/rustc_mir_transform/src/ffi_unwind_calls.rs (hunk 2 above)
Comment (nbdd0121) [reply compiler/rustc_mir_transform/src/ffi_unwind_calls.rs:49]: These lines are copied from `abort_unwinding_calls.rs` which uses the same checks.

File: compiler/rustc_metadata/src/dependency_format.rs (hunk 3 above)
Comment (nbdd0121) [reply compiler/rustc_metadata/src/dependency_format.rs:402]: Sorts of. Previously any crate can be linked to panic-abort; this line only has to be removed because the required panic strategy becomes three-value. You can think of the all previous `PanicStrategy::Unwind` maps to `None` in this PR.

File: library/std/src/lib.rs (hunk 4 above)
Comment (nbdd0121) [reply library/std/src/lib.rs:214]: I actually want to make it a default for compiling std and all its dependencies -- but not sure how.

File: compiler/rustc_interface/src/passes.rs (hunk 5 above)
Comment (nbdd0121) [reply compiler/rustc_interface/src/passes.rs:957]: Yes. This query uses `mir_built`, which will be stolen later. So I followed the same pattern as `check_unsafety` (which also uses `mir_built`, by making sure it's evaluated before a query that could steal it.

File: compiler/rustc_mir_transform/src/ffi_unwind_calls.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_mir_transform/src/ffi_unwind_calls.rs:49]: Hm, not sure how that makes this any less of a concern.^^ If anything, duplicating such subtle logic makes it worse...

Comment (nbdd0121) [reply compiler/rustc_mir_transform/src/ffi_unwind_calls.rs]: Just realized that we have `DefKind::is_fn_like`. Replaced the match in both files with that.

File: compiler/rustc_mir_transform/src/abort_unwinding_calls.rs [hunk 6]
```diff
@@ -31,11 +31,7 @@ impl<'tcx> MirPass<'tcx> for AbortUnwindingCalls {
 
         // We don't simplify the MIR of constants at this time because that
         // namely results in a cyclic query when we call `tcx.type_of` below.
-        let is_function = match kind {
-            DefKind::Fn | DefKind::AssocFn | DefKind::Ctor(..) => true,
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/abort_unwinding_calls.rs:35]: This is a change in behavior; `is_fn_like` returns `false` for a `Ctor`.

Comment (nbdd0121) [reply compiler/rustc_mir_transform/src/abort_unwinding_calls.rs:35]: It's fine; ctor won't call other functions.

**Comment by RalfJung** [comment]
> If anywhere in the crate graph is a panic=abort crate, then all call sites where foreign exceptions could enter the Rust part of the program are in a panic=abort crate, and hence will lead to immediate abort. These call sites are C-unwind call sites to FFI functions or function pointers -- a call site calling a Rust function cannot be where anything enters Rust.
> 
> There will also be no exceptions generated from Rust itself, and hence it is correct that a "Rust" ABI function can never unwind.

If :point_up: is indeed the key invariant, I think it should be explicitly documented as a comment somewhere in the part of the code that is responsible for upholding it.

File: compiler/rustc_mir_transform/src/ffi_unwind_calls.rs [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        if tcx.has_ffi_unwind_calls(def_id) {
+            // Given that this crate is compiled in `-C panic=unwind`, the `AbortUnwindingCalls`
+            // MIR pass will not be run on FFI-unwind call sites, therefore a foreign exception
+            // can enter Rust through these sites.
+            //
+            // On the other hand, crates compiled with `-C panic=abort` expects that all Rust
+            // functions cannot unwind (whether it's caused by Rust panic or foreign exception),
+            // and this expectation mismatch can cause unsoundness (#96926).
+            //
+            // To address this issue, we enforce that if FFI-unwind calls is used in a crate
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/ffi_unwind_calls.rs:147]:
```suggestion
            // To address this issue, we enforce that if FFI-unwind calls are used in a crate
```

File: compiler/rustc_mir_transform/src/ffi_unwind_calls.rs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+            // compiled with `panic=unwind`, then the final panic strategy must be `panic=unwind`.
+            // This will ensure that no crates will have wrong unwindability assumption.
+            //
+            // It should be noted that it is okay to link `panic=unwind` into a `panic=abort`
+            // program if it contains no FFI-unwind calls. In such case foreign exception can only
+            // enter Rust in a `panic=abort` crate, which will lead to an abort. There will also
+            // be no exceptions generated from Rust, so the assumption which `panic=abort` crates
+            // hold, that no Rust function can unwind, indeed holds for crates compiled with
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/ffi_unwind_calls.rs:155]:
```suggestion
            // make, that no Rust function can unwind, indeed holds for crates compiled with
```

**Comment by jackh726** [comment]
@RalfJung @Amanieu do one of you want to take over officially as reviewer for this? I'm not a good reviewer here and not sure rerolling is the best action.

**Comment by RalfJung** [comment]
I can't really speak to the implementation of this, just the general soundness of the argument. The last commit adds a comment that makes me happy.
So r? @Amanieu I hope :)

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang/rust/pull/98605#issuecomment-1168495819)

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang/rust/pull/98667#issuecomment-1169912321)

@bors r-

**Comment by nbdd0121** [comment]
Currently the lint only fires when `-C panic=unwind` is used -- if `-C panic=abort` is used then the lint does not fire.

It's possible to have `ffi_unwind_calls` lint to fire regardless current panic statregy -- but doing so would require me to change `fn_can_unwind` from returning bool to returning a tristate (`CanUnwind`, `NoUnwind`, `FollowPanicStrategy`). If that behaviour is desirable I can do it in a follow up PR.

File: library/std/src/lib.rs (hunk 4 above)
Comment (Mark-Simulacrum) [reply library/std/src/lib.rs:214]: @nbdd0121 was this followed up on? In the future, please feel free to reach out to #t-infra on Zulip for questions like this, we can help with the instrumentation necessary. In this case you'd likely add to the deny'd lints in bootstrap, around [here](https://github.com/rust-lang/rust/blob/53792b9c5c3cded466a073e4b9a9e51d0007b9ab/src/bootstrap/builder.rs#L1885), with a gate on Mode::Std.

**Comment by Mark-Simulacrum** [comment]
@nbdd0121 @Amanieu was a perf regression expected here? Do you have ideas on what caused it? It seems to me that it's pretty consistently a small regression across multiple benchmarks, so I suspect it's worth following up on to try and see if we can eliminate it or at least understand the cause.

File: library/std/src/lib.rs (hunk 4 above)
Comment (nbdd0121) [reply library/std/src/lib.rs:214]: I created a [thread](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Enable.20a.20lint.20when.20compiling.20std.20dependencies) in t-libs. The lint should be denied for std and its dependencies but not denied for panic runtimes, and it seems that this isn't quite achievement at the moment.
